drivers: ioapic: fix the cast for the offset

enjiamai opened the pull request:
The cast for the offset variable will make incorrect register
 access when it is over 0x80. Change the char cast to unsigned
 char to fix it.

Fixes #49803.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

stephanosio commented on drivers/interrupt_controller/intc_ioapic.c:
```diff
@@ -438,7 +438,7 @@ static uint32_t __IoApicGet(int32_t offset)
 	key = irq_lock();
 
-	*((volatile uint32_t *) (IOAPIC_REG + IOAPIC_IND)) = (char)offset;
+	*((volatile uint32_t *) (IOAPIC_REG + IOAPIC_IND)) = offset;
```
If any, it should be casting to `unsigned char` here since the upper 24 bits of the `IOREGSEL` register are reserved (must be 0).

enjiamai replied:
Thanks for the review, Got it fixed!

stephanosio commented on drivers/interrupt_controller/intc_ioapic.c:
```diff
@@ -108,8 +108,8 @@ __pinned_bss
 
-static uint32_t __IoApicGet(int32_t offset);
-static void __IoApicSet(int32_t offset, uint32_t value);
+static uint32_t __IoApicGet(uint32_t offset);
+static void __IoApicSet(uint32_t offset, uint32_t value);
```
all these `int32_t` to `uint32_t` type changes seem unnecessary.

enjiamai replied:
Ok, recovery the change here.

enjiamai committed:
drivers: ioapic: fix the cast for the offset

The cast for the offset variable will make incorrect register
access when it is over 0x80. Change the char cast to unsigned
char to fix it.

Fixes #49803.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

andyross reviewed (commented):
Style whine, not worth a -1 but sort of a code smell thing:

This value is being assigned through a uint32_t pointer.  Why is the cast even there?  If there are high bits set in the "offset" value that should be masked off, then we should be explicit about that.  If there are not, casting to a type that doesn't match the rvalue seems like a bug to me (and it was!).

enjiamai commented:
> This value is being assigned through a uint32_t pointer. Why is the cast even there? If there are high bits set in the "offset" value that should be masked off, then we should be explicit about that. If there are not, casting to a type that doesn't match the rvalue seems like a bug to me (and it was!).

Thanks for the review! @andyross  . I was also thinking to make the offset to uint32, will do it later.
